provider/coredns/coredns.go

```diff
@@ -148,10 +177,53 @@ func (c etcdClient) DeleteService(ctx context.Context, key string) error {
 	ctx, cancel := context.WithTimeout(ctx, etcdTimeout)
 	defer cancel()
 
+	if owned, err := c.isOwnedBy(ctx, key); err != nil {
+		return err
+	} else if !owned {
+		return fmt.Errorf("key %q is not owned by this service", key)
+	}
+
 	_, err := c.client.Delete(ctx, key, etcdcv3.WithPrefix())
 	return err
 }
 
+func (c etcdClient) isOwnedBy(ctx context.Context, key string) (bool, error) {
+	if !c.strictlyOwned {
+		return true, nil
+	}
+
+	r, err := c.client.Get(ctx, key)
+	switch {
+	case err != nil:
+		return false, err
+	case r == nil:
+		return true, nil
+	case len(r.Kvs) > 1:
+		return false, fmt.Errorf("unexpected case, etcd should never return multiple results by a key")
+	case len(r.Kvs) == 0:
+		return true, nil
+	}
+	for _, n := range r.Kvs {
+		svc, err := c.unmarshalService(n)
+		if err != nil {
+			return false, err
+		}
+
+		if svc.OwnedBy == c.ownerID {
+			return true, nil
+		}
+	}
+	return false, nil
```

Comment:
```suggestion
	 svc, err := c.unmarshalService(r.Kvs[0])
 if err != nil {
  return false, fmt.Errorf("failed to unmarshal value for key %q: %w", key, err)
 }

 return svc.OwnedBy == c.ownerID, nil
```

As we passing full key, not sure how is possible to return 2 results